### ISSUE TITLE
Socket.Receive fails on unbound UDP socket when tracing is enabled

### ISSUE THREAD

**Issue by frankracis-work** [dotnet/runtime#43521]
### Description
Calling `Socket.Receive` on a UDP client can cause an exception to be thrown when tracing is enabled.

Source to reproduce:

```c#
using System.Diagnostics;
using System.Diagnostics.Tracing;
using System.Net;
using System.Net.Sockets;

namespace UdpCrash
{
	class MyListener : EventListener
	{
		protected override void OnEventSourceCreated(EventSource eventSource)
		{
			base.OnEventSourceCreated(eventSource);
			if (eventSource.Name == "Microsoft-System-Net-Sockets")
			{
				EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.All);
			}
		}
	}

	class Program
	{
		static void Main(string[] args)
		{
			// Comment this line out and it works
			new MyListener();
			// DNS Query for www.microsoft.com
			var query= new byte[] {0x68,0xa3,0x01,0x00,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x03,0x77,0x77,0x77,
				0x09,0x6d,0x69,0x63,0x72,0x6f,0x73,0x6f,0x66,0x74,0x03,0x63,0x6f,0x6d,0x00,0x00,
				0x01,0x00,0x01};
			var udp = new UdpClient();
			var remote = new IPEndPoint(IPAddress.Parse("8.8.8.8"), 53);
			udp.Client.SendTo(query, SocketFlags.None, remote);
			var buffer = new byte[512];
			var rxCount = udp.Client.Receive(buffer);
			Debug.Assert(rxCount > 10);
			Debug.Assert(buffer[0] == query[0]);
			Debug.Assert(buffer[1] == query[1]);
		}
	}
}
```

On Windows, get the exception
```
Unhandled exception. System.Net.Sockets.SocketException (10057): A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied.
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.get_RemoteEndPoint()
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, SocketError& errorCode)
   at System.Net.Sockets.Socket.Receive(Byte[] buffer)
   at UdpCrash.Program.Main(String[] args) in C:\v16projects\UdpCrash\UdpCrash\Program.cs:line 34
```
On Linux, it's
```
Unhandled exception. System.Net.Sockets.SocketException (107): Transport endpoint is not connected
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.get_RemoteEndPoint()
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, SocketError& errorCode)
   at System.Net.Sockets.Socket.Receive(Byte[] buffer)
   at UdpCrash.Program.Main(String[] args) in /home/racis/src/UdpCrash/UdpCrash/Program.cs:line 34
```

The offending line is here:
https://github.com/dotnet/runtime/blob/149964acb7dd9bbeaa270a755aa8a80fadfc0ef1/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1424

Accessing `RemoteEndPoint` on an unbound socket causes an exception to be thrown.   The trace message should be something like `$"SRC:{LocalEndPoint} DST:{(Connected?RemoteEndPoint.ToString():"(none)")} size:{size}"`

### Configuration
.NET Core 3.1
Windows and Ubuntu Linux, x64

### Other information
Because the condition check is only for enabled/disabled, not the severity+keyword, the only workaround is to completely disable socket event tracing.

**Comment by wfurt** [comment]
do you want to prepare PR for 6.0 @frankracis-work ? The proposed change makes sense to me.

**Comment by stephentoub** [comment]
> do you want to prepare PR for 6.0

Does this repro on .NET 5 / master? e.g. this throws on .NET Core 3.1:
```C#
using System;
using System.Net;
using System.Net.Sockets;

var query= new byte[] { 0x68,0xa3,0x01,0x00,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x03,0x77,0x77,0x77,
                        0x09,0x6d,0x69,0x63,0x72,0x6f,0x73,0x6f,0x66,0x74,0x03,0x63,0x6f,0x6d,0x00,0x00,
                        0x01,0x00,0x01};
var udp = new UdpClient();
udp.Client.SendTo(query, SocketFlags.None, new IPEndPoint(IPAddress.Parse("8.8.8.8"), 53));
_ = udp.Client.RemoteEndPoint;
```
but doesn't throw (with RemoteEndPoint returning null) on .NET 5.

**Comment by wfurt** [comment]
should we close it then? I don't think this would meet servicing bar.

**Comment by frankracis-work** [comment]
Here was the line that fixed it:
https://github.com/dotnet/runtime/pull/34986/files#r408977642
I can confirm that this fixed the tracing issue as well.

**Comment by karelz** [comment]
Verified that it works on 5.0, but fails on 3.1. Marking as fixed in 5.0.
